### PR TITLE
[API-2092] Add packages for defining and bundling workers

### PR DESCRIPTION
**Pull request by danschultz**
## Summary

Adds packages for defining and bundling web workers. Supports creating separate bundles for workers, loading worker modules, and automatically finding `Transferables` to pass to/from workers.

https://vertexvis.atlassian.net/browse/API-2092

### Defining a Web Worker

Create a JS/TS file and wrap your worker implementation with `defineWorker`.

```ts
// ./worker.ts
import { defineWorker } from '@vertexvis/web-workers';

export type AddFn = (a: number, b: number) => Promise<number>;

const add: AddFn = async (a, b) => {
  return a + b;
}

defineWorker(add);
```

### Loading and Spawning a Web Worker

Use the `worker:` path prefix to bundle a web worker. The Rollup plugin will
transform the import and expose a `loadWorker` function to load the bundled
worker module. The worker module exports functions for spawning workers and
worker pools.

```ts
// ./main.ts
import type { AddFn } from './worker';
import { loadWorker } from 'worker:./worker';

async function main(): Promise<void> {
  const { spawnWorker, makeController } = await loadWorker<AddFn>();

  const controller = makeController(); // optional controller to stop worker.
  const add = await spawnWorker(controller);
  const sum = await add(1, 2)
  console.log('sum', sum); // 3

  controller.terminate();
}
```

### Worker Pools

Spawn worker pools to create a set of workers that pull work from a queue.

```ts
import type { AddFn } from './add';
import { loadWorker } from 'worker:./add';

async function main(): Promise<void> {
  const { spawnPool, makeController } = await loadWorker<AddFn>();

  const controller = makeController();
  const pool = await spawnPool({ controller });
  pool
    .queue((sum) => add(1, 2))
    .then((sum) => console.log('sum', sum)); // 3

  controller.terminate();
}
```

`pool.queue()` will return a promise that will resolve with the task's result.
Use `Promise.then()` to handle the result vs using `await` when scheduling work
within a loop. This is because `await` will block until the task completes.

You can customize the number of workers spawned and concurrency by passing in
options to `spawnPool`.

```ts
interface CreatePoolOptions {
  // The number of concurrent tasks to run per worker. Defaults to 1.
  concurrency?: number;

  // Max number of tasks to queue. If count is reached, an error is thrown.
  maxQueuedJobs?: number;

  // The name of the pool for debugging.
  name?: string;

  // The number of workers to spawn. Defaults to number of CPU cores.
  size?: number;

  // A controller to terminate the pool.
  controller?: WorkerController;
}
```

### Worker Termination

A `WorkerController` is used to terminate a worker. A worker module exports a
`makeController` function to create a controller. You can pass a controller to
multiple workers and pools to terminate them together.

```ts
import type { WorkerFn } from './worker1';
import { loadWorker as loadWorker1 } from 'worker:./worker1';

async function main(): Promise<void> {
  const { spawnWorker, makeController } = await loadWorker1<WorkerFn>();
  const { spawnPool } = await loadWorker2<WorkerFn>();

  const controller = makeController();

  const worker = await spawnWorker(controller);
  const pool = await spawnPool({ controller });

  // Terminate the worker and pool.
  await controller.terminate();
}
```

## Bundling Workers

```js
import resolve from '@rollup/plugin-node-resolve';
import typescript2 from 'rollup-plugin-typescript2';
import { terser } from 'rollup-plugin-terser';
import workers from '@vertexvis/rollup-plugin-web-workers';

export default {
  input: 'src/main.ts',
  output: {
    dir: './dist',
    format: 'esm',
  },
  plugins: [
    resolve(),
    workers({
      // Workers are bundled separately. Pass a list of plugins to use when
      // bundling the worker.
      plugins: [resolve(), typescript(), terser()]
    }),
    typescript2(),
  ]
}
```

### CODE REVIEW

File: packages/rollup-plugin-web-workers/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import { OutputChunk, Plugin, rollup } from 'rollup';
+
+interface ThreadsPluginOptions {
+  plugins?: Plugin[];
+}
+
+const workerPrefix = 'worker:';
+const workerModulePrefix = 'worker-module:';
+
+export default function ({ plugins = [] }: ThreadsPluginOptions): Plugin {
```
Comment (danschultz) [inline packages/rollup-plugin-web-workers/src/index.ts:11]: Rollup plugin to bundle workers into a module.

File: packages/web-workers/src/worker.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Defines a web worker implementation. This function wraps the `expose` method
+ * of Threads and handles automatic conversion of Transferable types.
+ *
+ * @param impl The worker implementation.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Transferable
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function defineWorker(
```
Comment (danschultz) [inline packages/web-workers/src/worker.ts:18]: Implementation to define web workers.

File: packages/web-workers/extensions.d.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/**
+ * Provides static analysis for import paths that are prefixed with `worker:`.
+ */
+declare module 'worker:*' {
```
Comment (danschultz) [inline packages/web-workers/extensions.d.ts:4]: Type defs for worker modules.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+## What
```
Comment (therockstorm) [inline .github/pull_request_template.md]: Should need this, it gets pulled in from the `.github` repo.

File: .nvmrc [hunk 5]
```diff
@@ -0,0 +1 @@
+v14.16.0
```
Comment (therockstorm) [inline .nvmrc]: v14.17 is latest LTS

**Review by therockstorm — APPROVED** [review]

**Pull request merged**